Add Docker container to support viewing HTML reports with visualizations

cvoltz opened the pull request:
## Description
* Add ``Dockerfile`` and scripts to build a [Docker](https://docs.docker.com/install/linux/docker-ce/centos/) container image and start and stop a Docker container so users can easily view the HTML reports, including those with visualizations.
* Add ``man`` pages for the new scripts.
* Update the **SEE ALSO** section of the existing ``man`` pages to reference the new scripts.
* Update the LinuxKI installation script to automatically build the container image if Docker is already installed. 

The Docker container exposes ports 80 and 443 but the SSL certificate is self-signed so the user must add an exception for the certificate if HTTPS is used. The ``kivis-start`` script starts a container with the exposed ports on random ports to prevent conflicts. It looks up the port mapped to the exposed port 80 and opens the default browser using that port so the user doesn't need to worry about which port is actually being used. The random ports which the exposed ports are mapped to can be discovered using the standard docker commands (e.g., ``docker ps``, ``docker inspect``, ``docker port``) but the ``kivis-start`` script outputs the port mapping.

The Docker container image and container which are managed by the ``kivis-*`` scripts are called ``linuxki``. This makes it easy for a user to use the scripts without having to specify the container image name or container name but it also prevents more than one container from being used at a time.  The ``kivis-start`` script will automatically stop any existing container before starting a new one. Containers are automatically removed when they are stopped.

It is also possible to create the Docker container image and create and destroy containers directly from the provided ``Dockerfile`` using Docker commands, for those who are familiar with them. The ``kivis-*`` commands are simple wrappers around the Docker commands for those who aren't familiar with Docker or who just want a simple and short command to view the HTML reports.

## Typical usage

### Build the container image
The container image only needs to be built once. If the container image was not built at install time, run:
```
kivis-build
```

### Generate the HTML report
Generate the HTML reports as usual. For example:
```
kiall -r -V
```

### View the report
To view the reports, from the directory containing the reports (or a parent directory), run:
```
kivis-start
```
The default browser will automatically be opened to the report if there is a single one in the current directory or its subdirectories. If there is more than one report, the browser will be opened to a page which provides a directory listing which the user can navigate to select the desired report.

When finished viewing the report, run:
```
kivis-stop
```

MarkCRay commented:
Hi Christopher,

Thanks for the contribution.   I think the feature will be helpful for those that are familiar with dockers already, and those that have graphical terminal so they can run a browser on their Linux server.    

I would prefer if the linuxki_install script was not modified.   If docker is installed and not setup properly, you can get some docker error when installing LinuxKI.   Its annoying you don't want to use the new docker feature, so I would prefer this to be totally optional.   Here's an example:

  Installing : linuxki-5.4b-1.noarch                                                                                                                                                                    1/1
        Finishing installation of the Linux KI Toolset ...
        Running depmod - this might take a moment or two ...
docker - version 1.5
Copyright 2003, Ben Jansens <ben@orodu.net>

Usage: docker [OPTIONS]

Options:
  -help             Show this help.
  -display DISLPAY  The X display to connect to.

cvoltz commented:
I made what I thought was the reasonable assumption that if the `docker` command was available, Docker was installed and configured properly. In this case, it looks like [another package](http://icculus.org/openbox/2/docker/) (also called `docker`) is installed which provides a command called `docker` which is unrelated to Docker. In Debian/Ubuntu, the Docker package is `docker.io` and not `docker`. Note that the `docker.io` package shouldn't be used since it is outdated. The instructions for installing Docker on [Debian](https://docs.docker.com/install/linux/docker-ce/debian/#set-up-the-repository) and [Ubuntu](https://docs.docker.com/install/linux/docker-ce/ubuntu/#set-up-the-repository) install the `docker-ce` package and they specifically document removing the older and potentially conflicting packages `docker`, `docker-engine`, and `docker.io`.

Generally, it is better for the user if things gets configured automatically, when possible, as it makes it easier for the user. I can see several options:

1. Hide the error message at installation.
I can update `kvis-build` to add an option to silently exit when an error occurs and then modify the changes to the `linuxki_install` script to use that argument. Then you won't see any error messages during the LinuxKI package installation but, if you run the command manually, you'll still get the error message you need to figure out why things aren't working. This would be a pretty quick change to make.

1. Require the user to manually run `kvis-build`.
If you really want (though I don't recommend it), I can update the patch to just remove the changes to `linuxki_installer` but then you will need to update the documentation to tell the user they need to manually run `kvis-build` before trying to run `kvis-start`. This is a trivial change to make.

1. Make `kvis-start` always call `kvis-build`.
I could update `kvis-start` to call `kvis-build` when it starts (so the user doesn't have to worry about running `kvis-build` and no changes would be required in `linuxki_installer`. The downsize of doing this, however, is that it will make `kvis-start` run slower so startup time will be increased. This seems like it should be a quick change to make but I doubt I will be happy with the startup time so I'll end up spending additional time trying to reduce the startup time so the result will be that it will take longer to implement this option.

Irrespective of which option you select, I could also add a check into `kvis-build` for the case where there is a conflict with the Debian/Ubuntu `docker` package or when Docker isn't installed so it will output a message noting that Docker isn't installed. This really shouldn't be something I have to add to the code since the documentation which specifies the Docker requirement should note this potential conflict, but it would be nice for a user who doesn't read the documentation.

Let me know what you want me to do.

MarkCRay commented:
Hi Christopher.    I would prefer to leave the docker stuff out of linuxki_install.   The docker package for visualization is totally optional.   My preference would be to add a check in kivis-start to see if the docker container is built, and if not to proceed to call kivis-build.   Or in kivis-start, you can try the "docker start", and if it fails, then do a kivis-build and retry the docker start.   That way the kivis-build would only be done on the first kivis-start.     The user could also optionally run kivis-build manually.

MarkCRay commented:
You mentioned earlier:

_This makes it easy for a user to use the scripts without having to specify the container image name or container name but it also prevents more than one container from being used at a time._

I'm concerned on a shared system that a user may execute kivis-stop or kivis-start (which executes kivis-stop) while the linuxki container is in use by another user.   

A couple more comments...

- fully qualify the kivis-* script names in the scripts.   While the /opt/linuxki  _should_ be in the PATH, using the full name would prevent any PATH problems.    Especially if you installed LinuxKI and haven't logged off and back on to pickup the PATH from the /etc/profile.d/linuxki.sh file.   
- update manpages to include note about proxy variables and selinux

Mark

cvoltz commented:
I updated the pull request with commits to address the things we talked about. I also added a check to ensure `docker` is from Docker and not from the `docker` package in Debian/Ubuntu. Take a look at it and let me know if there is anything else which needs to be fixed.

MarkCRay commented:
Hello Christopher.   Please remove the kivis-build from the linuxki_install script.    Also, in the linuxki-remove script please be sure the docker is stopped and the image is removed.

MarkCRay commented:
Also, in the Dockerfile, there is a reference to the RPM package - linuxki-5.4-2.noarch.rpm.   If this will need to be updated for each release, I can add something to automate this so it pulls the correct version.   Just let me know.

MarkCRay commented:
There is a releases page now.  See...

https://github.com/HewlettPackard/LinuxKI/releases
